### ISSUE TITLE
t.mail.smtp.ESMTPSenderFactory requireTransportSecurity fails to recognize when SSL is in use

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3989]
|<img alt="jiggle's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| jiggle reported|
|-|-|
|Trac ID|trac#3989|
|Type|defect|
|Created|2009-08-28 15:56:16Z|
|Branch|https://github.com/twisted/twisted/tree/esmtp-ssl-3989|

If one uses t.reactor.connectSSL() to establish an SSL connection to a SMTP server, and requireTransportSecurity=True, it still attempts to use TLS (over SSL) and usually fails (the server won't allow it).

It should recognize that TLS is already present...

Current work-around is to set requireTransportSecurity=False when creating an instance of the factory.

Failure in this case (from the server, with an SSL connection already present is):
Failure: twisted.mail.smtp.TLSRequiredError: 502 Server does not support secure communication via TLS / SSL

Attachments:

* [3989-ESMTP_SSL-draft1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7ce/7ce4b1ceb5d18c9cef085ae199ece90f8b19d124/bbebdf4fee0281904a885cccccec567f1aed8395.patch) (5659 bytes) - added by argonemyth on 2012-07-11 15:27:51Z - first attempt to fix the bug!
* [3989-ESMTP_SSL-draft2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7ce/7ce4b1ceb5d18c9cef085ae199ece90f8b19d124/fec07fb66a63730c6086579d631a4ecda2551431.patch) (5156 bytes) - added by argonemyth on 2012-08-08 12:30:56Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3989 3989
type__defect defect
reporter__jiggle jiggle
priority__normal normal
milestone__ 
branch__branches_esmtp_ssl_3989 branches/esmtp-ssl-3989
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ESMTP_SSL_TLS ESMTP SSL TLS
time__1251474976000000 1251474976000000
changetime__1356901208000000 1356901208000000
version__None None
owner__exarkun exarkun
cc__argonemyth
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)<a name="note_2"></a>|@argonemyth set owner to @argonemyth|
|-|-|

Going to make **connectSSL** work without manually change the default value of **requireTransportSecurity**.

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)<a name="note_3"></a>|@argonemyth removed owner|
|-|-|

Fixing the bug is fairly easy, as I followed exarkun's suggestions. But, I spent pretty long time to write the proper test case for it. 

Exarkun asked me to deprecate the **tlsMode** in **ESMTPClient**, but I am not sure how to deprecate a variable properly. Since the chance of someone setting the value to **True** is rare, I removed all the references of **tlsMode** in smtp.py. But, I don't think it's a good an idea to just remove them all, let me know the proper way of handling this.

As for the test case, I tried different approaches to simulate **reactor.connectSSL** (tried to use existing classes and helpers), but still, the file ended up with 4 new classes. Please review the draft 1 of the patch and let me know if you have better idea for the test case.

Thank you very much!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @argonemyth|
|-|-|

Thanks.  Two main comments:

  1. Don't delete `tlsMode`.  CompatibilityPolicy describes how we make incompatible changes.  This will involve deprecating the attribute, releasing that a couple times, and then removing it.  Please file a new ticket for introducting that deprecation.  For now, for this branch, leave the `tlsMode` attribute in place, but we don't have to use it anymore.
  1. The test says it is going to "simulate" `reactor.connectSSL`.  That sounds excellent.  However, what the test actually does is use the real `reactor.connectSSL`.  Any time it's possible to avoid creating _real_ TCP connections (SSL runs over TCP), we want to avoid it.  Real TCP connections are subject to too many external influences to be reliably used in unit tests.
    1. See `twisted.test.proto_helpers.MemoryReactor` for a bare-bones simulation of `connectSSL` that would help out with this test.  You'll probably have to write one or two more similar "memory" simulations to complete the test, though - for example, a simulation of the transport object (although we have `twisted/test/iosim.py`, it isn't documented and does some weird extra things that might not be helpful in this case).
    1. Also, test methods need docstrings too.

Please feel free to find me on IRC or post to twisted-python if you'd like to discuss how to approach either of these points.

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)<a name="note_6"></a>|@argonemyth removed owner|
|-|-|

Replying to [exarkun](#note_5):

Uploaded the second version of the patch...

>   1. Don't delete `tlsMode`.  CompatibilityPolicy describes how we make incompatible changes.  This will involve deprecating the attribute, releasing that a couple times, and then removing it.  Please file a new ticket for introducting that deprecation.  For now, for this branch, leave the `tlsMode` attribute in place, but we don't have to use it anymore.

Kept the `tlsMode` as the class attribute, but removed any other appearances of `tlsMode` in the class. Hope that's what you mean by "we don't have to use it anymore"!

>   2. The test says it is going to "simulate" `reactor.connectSSL`.  That sounds excellent.  However, what the test actually does is use the real `reactor.connectSSL`.  Any time it's possible to avoid creating _real_ TCP connections (SSL runs over TCP), we want to avoid it.  Real TCP connections are subject to too many external influences to be reliably used in unit tests.

Well, In the second patch, I used the `MemoryReactor`, but I made the `connectSSL` method returning a protocol object which connected with a fake SSL StringTransport. I don't know if it's a good solution, but it's the simplest I can come up with right now, and it did the job of testing `ESMTPClient.tryTLS` method.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [36672]) Branching to 'esmtp-ssl-3989'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks.

  1. The whole `MemoryReactor.connectSSL` process is unnecessary.  All you need to do is instantiate the protocol (`ESMTPClient`) being tested, connect it to a transport (call `makeConnection` on it), and then deliver data to it.
  1. There is some missing test coverage for some of the code paths through `tryTLS`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/esmtp-ssl-3989)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun removed owner|
|-|-|

I addressed those points, made the tests properly skip when SSL dependencies are unavailable, and added a news fragment.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_11"></a>|@tomprince set owner to @exarkun|
|-|-|

The behavior and use of ```tryTLS}} has changed, and isn't entirely clear. It would be good to add a docstring explaining what it does.

Please commit after adding a docstring for {{{tryTLS``` (and optionally ```esmtpState_serverConfig``` and ```esmtpState_starttls```).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [36695]) Merge esmtp-ssl-3989

Author: argonemyth, exarkun
Reviewer: tom.prince
Fixes: [#3989](https://github.com/twisted/twisted/issues/3989)

Teach `ESMTPClient` about running over SSL connections - specifically, that it does
not need to try to negotiate a new SSL session in this case.
